Support address_prefixes for multiple prefixes in subnet and virtualnetwork (ipv6)

fabianofranz opened the pull request:
The azure subnet resource supports addressPrefix when one prefix is available and addressPrefixes when two or or more are set (for instance, when configuring an ipv4 and ipv6 dual stack configuration for a network). Update azurerm_subnet to allow either address_prefix or address_prefixes as input, where the latter is a list of string. Check that at least one is set, and mark the older address_prefix as deprecated.

The virtual network resource allows inline subnets, but does not currently support this argument. When the address_prefixes argument is used, address_prefix will be null when terraform refreshes its state causing terraform to panic. This code updates also the virtual network resource to allow
address_prefixes to be used and for address_prefix to be null.

Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/5140
Takes over https://github.com/terraform-providers/terraform-provider-azurerm/pull/5139

abhinavdahiya commented:
maybe commits from https://github.com/terraform-providers/terraform-provider-azurerm/pull/6142 belong with this PR since they are addressing the same _subnets with multiple address prefixes_

fabianofranz commented:
@abhinavdahiya makes sense, I added commit https://github.com/terraform-providers/terraform-provider-azurerm/pull/6142 to this PR and closed the other one.

fabianofranz commented:
@katbyte PTAL

katbyte commented on azurerm/internal/services/network/resource_arm_virtual_network.go:
```diff
@@ -106,9 +106,15 @@ func resourceArmVirtualNetwork() *schema.Resource {
+						"address_prefixes": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
```
could we validate this?

fabianofranz commented:
@katbyte @mbfrahry I'm having a test failure with `ExactlyOneOf`, do you have any clue? Tks!

```
--- FAIL: TestProvider (0.13s)
    provider_test.go:12: err: 1 error occurred:
        	* resource azurerm_virtual_network: ExactlyOneOf: address_prefix references unknown attribute (address_prefix) at part (address_prefix)
```

mbfrahry commented:
Hey @fabianofranz, unfortunately  `ExactlyOneOf`, `ConflictsWith`, and the others have a limitation in that they don't work quite the way you want them to for Lists and Sets. The error you're seeing is because the `ExactlyOneOf` logic is looking for `address_prefix` but it should really be looking for `subnet.x.address_prefix` where `x` is the index of the set you want to compare. The tricky bit is that we don't know how many subnets there could be so we can't effectively check that `ExactlyOneOf` works. Unfortunately, we'll need to do some of this logic manually for now with a note to remove it when we release 3.0 of the provider in the future. 

For future uses of this, you can use it for cases where the List/Set has a `MaxItems: 1` because we know that `subnet.0.address_prefix` will exist. But we can't use that Validation for any List/Set that we don't know the size of.

fabianofranz commented on azurerm/internal/services/network/tests/resource_arm_virtual_network_test.go:
```diff
@@ -275,6 +296,25 @@ resource "azurerm_virtual_network" "test" {
+func testAccAzureRMVirtualNetwork_basic_addressPrefixes(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Since most already existing tests already use that, I'd suggest to handle this for all tests in a separate PR. ;-)

mbfrahry replied:
I agree we should do it for the rest but we should do the best practice now for any new features and we can fix up the rest when we can

fabianofranz commented:
> Hey @fabianofranz, unfortunately `ExactlyOneOf`, `ConflictsWith`, and the others have a limitation in that they don't work quite the way you want them to for Lists and Sets. The error you're seeing is because the `ExactlyOneOf` logic is looking for `address_prefix` but it should really be looking for `subnet.x.address_prefix` where `x` is the index of the set you want to compare. The tricky bit is that we don't know how many subnets there could be so we can't effectively check that `ExactlyOneOf` works. Unfortunately, we'll need to do some of this logic manually for now with a note to remove it when we release 3.0 of the provider in the future.
> 
> You can use it with List/Set that has a `MaxItems: 1` because we know that `subnet.0.address_prefix` will exist. But we can't use that Validation for any List/Set that we don't know the size of.
> 
> Also, I want to apologize for sending you down this rabbit hole. The files changed tab made it tricky to see that those values were in a Set. But the validation should be able to stay with the subnet resource

Ok @mbfrahry, thanks for the explanation. I updated this PR back to validations in the subnet resource and addressed most comments from you and @katbyte. I think it must be good for another round of reviews, thanks!

mbfrahry commented on azurerm/internal/services/network/tests/resource_arm_subnet_test.go:
```diff
@@ -442,6 +461,27 @@ resource "azurerm_subnet" "test" {
+func testAccAzureRMSubnet_basic_addressPrefixes(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
```suggestion
  name     = "acctestRG-n-%d"
```
Updating for best practices

mbfrahry commented on azurerm/internal/services/network/tests/data_source_virtual_network_test.go:
```diff
@@ -87,6 +110,30 @@ data "azurerm_virtual_network" "test" {
+func testAccDataSourceArmVirtualNetwork_basic_addressPrefixes(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctest%d-rg"
```
```suggestion
  name     = "acctest%d-n-rg"
```
Updating for best practices

mbfrahry commented on azurerm/internal/services/network/tests/data_source_subnet_test.go:
```diff
@@ -30,6 +30,28 @@ func TestAccDataSourceSubnet_basic(t *testing.T) {
 	})
 }
 
+func TestAccDataSourceAzureRMSubnet_basic_addressPrefixes(t *testing.T) {
```
Test now passes!

fabianofranz commented:
@mbfrahry thanks for the review, all latest comments were addressed.

> I am having issues getting the AllowMultipleAddressPrefixesOnSubnet feature for virtual networks registered

I'm facing exactly the same issue when trying to test it since yesterday, and I'm pretty sure it wasn't happening until a few days ago. Let's wait and see how your ticket goes. ;-)

So here goes [another PR](https://github.com/terraform-providers/terraform-provider-azurerm/pull/6493) that splits this one. If we're good to go with the subnets part of this, let's have the other PR going and once it lands, I'll rebase this one to focus on the VNet and wait for a solution about the test issue. Tks!

mbfrahry commented:
Thanks for splitting this out @fabianofranz, I just heard back from support and they said that this feature isn't available for the public use yet and they aren't approving people to use it. I'll move this PR to our Blocked milestone until they release this feature to the public.

abhinavdahiya commented:
Any updates on this @fabianofranz @mbfrahry

mbfrahry commented:
Closing this for now as it's not yet available for public use
